src/main/java/org/gaul/s3proxy/ShardedBlobStore.java

```diff
@@ -0,0 +1,677 @@
+        meta.put("s3proxy-sharded-superblock-shards",
+                Integer.toString(bucket.shards));
+        return meta.build();
+    }
+
+    private static String getShardContainer(ShardedBucket bucket, int shard) {
+        return String.format("%s-%d", bucket.prefix, shard);
+    }
+
+    private String getShard(String containerName, String blob) {
+        ShardedBucket bucket = buckets.get(containerName);
+        if (bucket == null) {
+            return containerName;
+        }
+        HashCode hash = SHARD_HASH.hashString(blob, StandardCharsets.UTF_8);
+        return ShardedBlobStore.getShardContainer(
+                bucket, Hashing.consistentHash(hash, bucket.shards));
+    }
+
+    private void checkSuperBlock(Blob blob, Map<String, String> expectedMeta,
+                                 String container) {
+        Map<String, String> currentSuperblockMeta =
+                blob.getMetadata().getUserMetadata();
+        for (String key : expectedMeta.keySet()) {
+            String current = currentSuperblockMeta.get(key);
+            String expected = expectedMeta.get(key);
+            if (!expected.equalsIgnoreCase(current)) {
+                throw new RuntimeException(String.format(
+                        "Superblock block for %s does not match: %s, %s",
+                        container, expected, current));
+            }
+        }
+    }
+
+    private boolean createShards(ShardedBucket bucket, Location location,
+                                 CreateContainerOptions options) {
+        Exception err = null;
+        ImmutableMap.Builder<Future<Boolean>, String> futuresMapBuilder =
+                new ImmutableMap.Builder<>();
+        ExecutorService executor = Executors.newFixedThreadPool(
+                Math.min(bucket.shards, MAX_SHARD_THREADS));
+        BlobStore blobStore = this.delegate();
+        IntStream.range(0, bucket.shards).forEach(
+            n -> futuresMapBuilder.put(executor.submit(
+                () -> blobStore.createContainerInLocation(
+                        location, ShardedBlobStore.getShardContainer(bucket, n),
+                        options)),
+                    ShardedBlobStore.getShardContainer(bucket, n))
+        );
+        executor.shutdown();
+        ImmutableMap<Future<Boolean>, String> futuresMap =
+                futuresMapBuilder.build();
+        boolean ret = true;
+        for (Future<Boolean> result : futuresMap.keySet()) {
+            String shard = futuresMap.get(result);
+            try {
+                ret &= result.get();
+            } catch (InterruptedException | ExecutionException e) {
+                logger.warn("Failed to create the shard {}: {}", shard, e);
+                if (err == null) {
+                    err = e;
+                }
+            }
+        }
+
+        if (err != null) {
+            throw new RuntimeException("Failed to create some shards", err);
+        }
```

Comment:
I think it's better to just throw the first exception instead of trying to log everything.  What value would 1,000 repeated log messages have?  This is repeated below.

Reply:
The idea was to allow identifying the specific failing shards, but your concern is valid. Switched to re-raising the first exception.